src/registry.rs

```diff
@@ -0,0 +1,243 @@
+    ///
+    /// - URL with scheme `file:///` or no scheme will return the default LocalFS store
+    /// - URL with scheme `s3://bucket/` will return the S3 store
+    /// - URL with scheme `hdfs://hostname:port/` will return the hdfs store
+    ///
+    /// If no [`ObjectStore`] found for the `url`, ad-hoc discovery may be executed depending on
+    /// the `url` and [`ObjectStoreRegistry`] implementation. An [`ObjectStore`] may be lazily
+    /// created and registered.
+    fn get_store(&self, url: &Url) -> Option<Arc<dyn ObjectStore>>;
+
+    /// List all registered URLs
+    fn list_urls(&self) -> Vec<String>;
+}
+
+/// The default [`ObjectStoreRegistry`]
+pub struct DefaultObjectStoreRegistry {
+    /// A map from scheme to object store that serve list / read operations for the store
+    object_stores: RwLock<HashMap<String, Arc<dyn ObjectStore>>>,
+}
+
+impl std::fmt::Debug for DefaultObjectStoreRegistry {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let stores = self.object_stores.read().unwrap();
+        f.debug_struct("DefaultObjectStoreRegistry")
+            .field("schemes", &stores.keys().cloned().collect::<Vec<_>>())
+            .finish()
+    }
+}
+
+impl Default for DefaultObjectStoreRegistry {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DefaultObjectStoreRegistry {
+    /// Create a new [`DefaultObjectStoreRegistry`] with no registered stores
+    pub fn new() -> Self {
+        let object_stores = RwLock::new(HashMap::new());
+        Self { object_stores }
+    }
+}
+
+///
+/// Stores are registered based on the scheme, host and port of the provided URL.
+///
+/// For example:
+///
+/// - `file:///my_path` will return the default LocalFS store
+/// - `s3://bucket/path` will return a store registered with `s3://bucket` if any
+/// - `hdfs://host:port/path` will return a store registered with `hdfs://host:port` if any
+impl ObjectStoreRegistry for DefaultObjectStoreRegistry {
+    fn register_store(
+        &self,
+        url: &Url,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>> {
+        let s = get_url_key(url);
+        let mut stores = self.object_stores.write().unwrap();
+        stores.insert(s, store)
+    }
+
+    fn get_store(&self, url: &Url) -> Option<Arc<dyn ObjectStore>> {
+        let s = get_url_key(url);
+        let mut stores = self.object_stores.write().unwrap();
+
+        if let Some(store) = stores.get(&s) {
+            return Some(Arc::clone(store));
+        }
+
+        match parse_url(url) {
+            Ok((store, _)) => {
+                let store = Arc::from(store);
+                stores.insert(s, Arc::clone(&store));
+                Some(store)
+            }
+            Err(_) => None,
+        }
+    }
+
+    fn list_urls(&self) -> Vec<String> {
+        let stores = self.object_stores.read().unwrap();
+        stores.keys().cloned().collect()
```

Comment:
This doesn't return URLs, but I believe host and port

Reply:
IIUC, the keys in `object_stores` are the returns from `get_url_key`. `get_url_key` currently returns a URL:

```rust
    format!(
        "{}://{}",
        url.scheme(),
        &url[url::Position::BeforeHost..url::Position::AfterPort],
    )
```

Reply:
Switched to return `Vec<Url>`